src/sentry/event_manager.py

```diff
@@ -1664,7 +1669,14 @@ def _save_aggregate_new(
     maybe_run_background_grouping(project, job)
 
     record_hash_calculation_metrics(
-        primary_grouping_config, primary_hashes, secondary_grouping_config, secondary_hashes
+        # Cast in lieu of a non-null assertion operator, which Python doesn't have
+        #
+        # TODO: The typing and necessity of casting here is gross, but might be able to be improved
+        # once hierarchical grouping is gone
+        cast(GroupingConfig, primary.config),
+        cast(CalculatedHashes, primary.hashes),
```

Comment:
Is this related to the primary return sig being `tuple[GroupingConfig, CalculatedHashes]`, while the wrapper function is `tuple[GroupingConfig | None, CalculatedHashes | None]`?

Reply:
Yeah - it's because the primary for sure returns data but the secondary might or might not, so I had to make the wrapper include the `None` possibility, even though I know then it's wrapping the primary, the `None` option doesn't apply. The casting was the least bad way I could come up with to handle that. If you have a better idea, I'm all ears.

(Also, I'll admit that I now can't remember why I thought that getting rid of hierarchical grouping might fix it. Maybe I just like to blame it for everything.)

Reply:
UPDATE: I thought of a better way. See https://github.com/getsentry/sentry/pull/65372.